### PR TITLE
Support references to outer type params for assoc consts

### LINKED ISSUE #28809: associated constants should support references to outer type parameters
associated constants should support references to outer type parameters

Here is an example, adapted from the `Zeroable` trait from `core::non_zero` ([playpen](http://is.gd/hf1VIN)):

```rust
#![feature(associated_consts)]
trait Zble { const Z: Self; }
impl Zble for i32 { const Z: i32 = 0; }

impl<T> Zble for *const T { const Z: *const T = ::std::ptr::null(); }

fn main() { }
```

An attempt to compile this yields the error:

```
<anon>:5:45: 5:46 error: cannot use an outer type parameter in this context [E0402]
<anon>:5 impl<T> Zble for *const T { const Z: *const T = ::std::ptr::null(); }
                                                     ^
<anon>:5:45: 5:46 error: use of undeclared type name `T` [E0412]
<anon>:5 impl<T> Zble for *const T { const Z: *const T = ::std::ptr::null(); }
                                                     ^
```

### PR DESCRIPTION
**Pull request by nagisa**
Fixes #28809 

r? @eddyb

### CODE REVIEW

File: src/test/run-pass/associated-const-outer-ty-refs.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+#![feature(associated_consts)]
+
+trait Lattice {
+    const BOTTOM: Self;
+}
+
+impl<T: 'static> Lattice for Option<T> {
```
Comment (eddyb) [inline src/test/run-pass/associated-const-outer-ty-refs.rs]: Maybe open an issue and add a FIXME here? We should support this.

**Comment by eddyb** [comment]
r=me with the FIXME added

**Comment by eddyb** [comment]
Nominated for beta because we might want to use this at stage0.

**Comment by nagisa** [comment]
I never ended using the features so there’s no need to backport to beta anymore.

**Pull request merged**